### PR TITLE
Task/HW-54195 - Introduce transfer method configuration repository

### PR DESCRIPTION
**Pull request by gmeyer-hw**
HW-54195 - Introduce the Transfer Method Configuration Repository Pattern:
- [x]  The Core-SDK task/HW-54195-Introduce-transfer-method-configuration-repository must be merged first
- Created TransferMethodRepository module;
- Added TransferMethodConfigurationRepositor and TransferMethodConfigurationRepositoryFactory;
- Refactoring the SelectTransferMethodTypePresenter to use Repository pattern;
- Introduce the GenericCellConfiguration to display HyperwalletCountry and HyperwalletCurrency as a cell.

### CODE REVIEW

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 1]
```diff
@@ -74,43 +82,39 @@ final class SelectTransferMethodTypePresenter {
...
-
-            guard let currencies = self.loadTranferMethodConfigurationCurrencies(for: selectedCountry) else {
-                view.showAlert(message: String(format: "no_currency_available_error_message".localized(),
-                                               selectedCountry.localized()))
-                return
+        transferMethodConfigurationRepository.getKeys(completion: self.getKeysHandler({ (result) in
+            if index == 0 {
+                self.showSelectCountryView(result?.countries())
+            } else {
+                guard !self.selectedCountry.isEmpty else {
```
Comment (jsaini-hw) [inline Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: no need of this guard statement

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 2]
```diff
@@ -146,44 +158,34 @@ final class SelectTransferMethodTypePresenter {
...
+            case .success(let keyResult):
+                successHandler(keyResult)
             }
+        }
     }
 
     /// Shows the Select Country View
-    private func showSelectCountryView(_ countries: [CountryCurrencyCellConfiguration]) {
+    private func showSelectCountryView(_ countries: [GenericCellConfiguration]?) {
+        guard let countries = countries else {
```
Comment (jsaini-hw) [inline Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: No need of this guard statement. Convert it to if

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 3]
```diff
@@ -192,51 +194,62 @@ final class SelectTransferMethodTypePresenter {
     }
 
     /// Shows the Select Currency View
-    private func showSelectCurrencyView(_ currencies: [CountryCurrencyCellConfiguration]) {
+    private func showSelectCurrencyView(_ currencies: [GenericCellConfiguration]?) {
+        guard let currencies = currencies  else {
```
Comment (jsaini-hw) [inline Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: No need of this guard statement. Convert it to if

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 4]
```diff
@@ -251,44 +264,41 @@ final class SelectTransferMethodTypePresenter {
...
-                view.showAlert(message: String(format: "no_currency_available_error_message".localized(),
-                                               country.localized()))
-                selectedCurrency = ""
-                view.countryCurrencyTableViewReloadData()
-                return
+    private func loadSelectedCurrency(_ currencies: [HyperwalletCurrency]?) {
+        guard let firstCurrency = currencies?.min(by: { $0.name < $1.name }) else {
+            view.showAlert(message: String(format: "no_currency_available_error_message".localized(),
+                                           selectedCountry.localized()))
+            selectedCurrency = ""
```
Comment (jsaini-hw) [inline Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: showAlert will anyways go back to previous UI so this code will never be executed

File: Sources/TransferMethod/SelectTransferMethodTypeTableViewController.swift [hunk 5]
```diff
@@ -31,6 +31,17 @@ public final class SelectTransferMethodTypeTableViewController: UITableViewContr
...
+    private var forceUpdate: Bool
+
+    init(forceUpdate: Bool) {
+        self.forceUpdate = forceUpdate
+        super.init(style: .plain)
+    }
+
+    //swiftlint:disable unavailable_function
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (jsaini-hw) [inline Sources/TransferMethod/SelectTransferMethodTypeTableViewController.swift]: add super.init(coder: aDecoder) instead of fatalError

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 6]
```diff
@@ -46,14 +50,18 @@ final class SelectTransferMethodTypePresenter {
...
-    private (set) var sectionData = [HyperwalletTransferMethodType]()
     private (set) var countryCurrencySectionData = [String]()
     private (set) var selectedCountry = ""
     private (set) var selectedCurrency = ""
 
+    private var transferMethodConfigurationRepository: TransferMethodConfigurationRepository {
+        return TransferMethodRepositoryFactory.shared.transferMethodConfigurationRepository()
+    }
+
+    var sectionData = [HyperwalletTransferMethodType]()
```
[the remark this answers is not in the document]
Comment (gmeyer-hw) [reply Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: sure

File: Sources/HyperwalletUI.swift [hunk 7]
```diff
@@ -68,9 +68,11 @@ public final class HyperwalletUI {
     /// Lists all transfer method types available based on the country, currency and profile type to create a new
     /// transfer method (bank account, bank card, PayPal account, prepaid card, paper check).
     ///
+    /// - Parameter forceUpdateDataManager: Forces to refresh the data manager
     /// - Returns: An instance of `SelectTransferMethodTypeTableViewController`
-    public func selectTransferMethodTypeTableViewController() -> SelectTransferMethodTypeTableViewController {
-        return SelectTransferMethodTypeTableViewController()
+    public func selectTransferMethodTypeTableViewController(forceUpdateDataManager: Bool = false)
```
[the remark this answers is not in the document]
Comment (gmeyer-hw) [reply Sources/HyperwalletUI.swift]: Sure

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 8]
```diff
@@ -125,9 +129,17 @@ final class SelectTransferMethodTypePresenter {
 
             strongSelf.user = result
 
-            Hyperwallet.shared.retrieveTransferMethodConfigurationKeys(
-                request: HyperwalletTransferMethodConfigurationKeysQuery(),
-                completion: strongSelf.transferMethodConfigurationKeyResultHandler())
+            DispatchQueue.main.async {
```
[the remark this answers is not in the document]
Comment (gmeyer-hw) [reply Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: We need to keep it until the UserRepository being implemented.

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 9]
```diff
@@ -146,44 +158,34 @@ final class SelectTransferMethodTypePresenter {
...
-    private func loadCountry() {
-        guard let countries = loadTranferMethodConfigurationCountries() else {
-            view.showAlert(message: "no_country_available_error_message".localized())
-            return
-        }
+    private func getKeysHandler(
+        _ successHandler: @escaping ((HyperwalletTransferMethodConfigurationKey?) -> Void),
+        _ failureHandler: (() -> Void)? = nil)
+        -> (Result<HyperwalletTransferMethodConfigurationKey?, HyperwalletErrorType>) -> Void {
+        return { [weak self] (result) in
```
[the remark this answers is not in the document]
Comment (gmeyer-hw) [reply Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: I'm going to remove the `[weak self]` check from the repository layer and keep it the presenter check.

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 10]
```diff
@@ -125,118 +121,109 @@ final class SelectTransferMethodTypePresenter {
...
-            }
-            strongSelf.selectedCountry = country.value
-            strongSelf.loadCurrency(for: country.value)
-            strongSelf.loadTransferMethodTypes(country: strongSelf.selectedCountry,
-                                               currency: strongSelf.selectedCurrency)
+        return { (country) in
+            self.selectedCountry = country.value
+            self.transferMethodConfigurationRepository
+                .getKeys(completion: self.getKeysHandler(success: { (result) in
+                    self.loadCurrencyAndTransferMethodTypes(result)
```
Comment (jsaini-hw) [inline Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: same as above

File: Sources/TransferMethod/SelectTransferMethodTypePresenter.swift [hunk 11]
```diff
@@ -251,44 +238,37 @@ final class SelectTransferMethodTypePresenter {
...
-            .currencies(from: country)?.min(by: { $0.name < $1.name })
-            else {
-                view.showAlert(message: String(format: "no_currency_available_error_message".localized(),
-                                               country.localized()))
-                selectedCurrency = ""
-                view.countryCurrencyTableViewReloadData()
-                return
+    private func loadCurrencyAndTransferMethodTypes(_ keys: HyperwalletTransferMethodConfigurationKey?) {
+        guard let firstCurrency = keys?.currencies(from: selectedCountry)?
+            .min(by: { $0.name < $1.name }) else {
```
Comment (jsaini-hw) [inline Sources/TransferMethod/SelectTransferMethodTypePresenter.swift]: no need of .min(by: { $0.name < $1.name }), just use first

File: UITests/Receipts/TransactionDetailTests.swift [hunk 12]
```diff
@@ -0,0 +1,198 @@
+import XCTest
```
Comment (jsaini-hw) [inline UITests/Receipts/TransactionDetailTests.swift]: This Test has been deleted from development branch. Why is this here?

Comment (gmeyer-hw) [reply UITests/Receipts/TransactionDetailTests.swift]: I removed this file.

File: Cartfile [hunk 13]
```diff
@@ -1 +1 @@
-github "hyperwallet/hyperwallet-ios-sdk" "1.0.0-beta03"
+github "hyperwallet/hyperwallet-ios-sdk" "Development"
```
Comment (jsaini-hw) [inline Cartfile]: development

File: TransferMethodRepository/Sources/TransferMethodConfigurationRepository.swift [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        _ transferMethodType: String,
+        _ transferMethodProfileType: String,
+        completion: @escaping (Result<HyperwalletTransferMethodConfigurationField?, HyperwalletErrorType>) -> Void) {
+        let fieldsQuery = HyperwalletTransferMethodConfigurationFieldQuery(
+            country: country,
+            currency: currency,
+            transferMethodType: transferMethodType,
+            profile: transferMethodProfileType
+        )
+        guard let transferMethodConfigurationFields = transferMethodConfigurationFieldsDic[fieldsQuery] else {
```
Comment (jsaini-hw) [inline TransferMethodRepository/Sources/TransferMethodConfigurationRepository.swift]: rename transferMethodConfigurationFieldsDic tp transferMethodConfigurationFieldsDictionary
